# Coding Session 05: Power and Null Results

**Theory → worked examples → all exercises at the end.**
Follow one canvassing case from turnout estimates to power and cost.
In Colab, save a copy to Drive, then run the cells in order.

## 1. The class case: the canvassing pilot

You manage a gubernatorial campaign. The field director randomly assigned
400 households in one district to a canvass attempt or control:

| | Control | Canvass |
|---|---:|---:|
| Households | 200 | 200 |
| Turnout | 42.0% | 43.5% |

The estimated effect is **+1.5 percentage points (pp)**, with a regression
**p-value of 0.76**. She recommends reallocating to digital ads.
Previous research summarized in the case suggests canvassing raises
turnout by about **1–3 percentage points per household assigned**.
The pilot needs to be precise enough to distinguish those effects from zero.

*Example: Week 5 case, pp. 1–2.*

### The theory from class

- **Absence of evidence is not evidence of absence.** “We could not detect
  an effect” does not establish “the effect is zero.”
- **Power:** if a specified effect really exists, how often would this
  experiment detect it? The 400-household example has about **6%
  power for +2 pp**: roughly 94 in 100 runs miss it, a **false negative**.
  A false positive instead detects an effect when none exists.
- Power rises with **sample size** and **effect size**, and falls with
  **noise**. Binary outcomes are most variable near 50% turnout.
  Choose the assumed effect from prior evidence, not the pilot's estimate.
- **MDE:** the smallest true effect detected with a chosen probability,
  usually **80%**. The pilot's MDE is about **14 pp**; around **20,000
  households** are needed to detect +2 pp with about 80% power.

“Well-powered” always means **powered for a particular effect**.
A more sensitive study's null is more informative about that effect;
neither a high p-value nor an MDE proves zero.

### The bounce, the interval, and the detection threshold

The **standard error (SE)** is how much the estimate would bounce across
repeated experiments. For the pilot it is about **5 pp**.

| Class rule | Meaning |
|---|---|
| 95% CI ≈ estimate ± 1.96 × SE | Effect sizes consistent with the data under the model. |
| Four times the sample → half the SE | With similar turnout rates and the same design, the CI is half as wide. |
| MDE ≈ 2.8 × SE (roughly 3 × SE) | 80% power with a two-sided 5% test. |
| Significance requires about 1.96 × SE | An observed estimate can be significant below the MDE. |

**Truth inflation:** when an underpowered study produces a significant
result, selection for significance tends to pick an unusually large estimate.
It does not make every small-study estimate too large.

## 2. Read the pilot

The course CSV has 400 household records: `household_id` is the ID,
`treatment` marks assignment (1/0), and `turned_out` is the case's binary
turnout outcome. It is teaching case data with one outcome per household,
not a record of completed conversations. Load it directly from the course
URL below; an internet connection is required.

In [1]:
# Import the tools we need. The names after "as" are short aliases:
# pd works with tables, np handles numerical calculations and random draws,
# and smf fits regressions using formulas written as strings.
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# Store the address of the course CSV as a string.
# Inside parentheses, Python joins adjacent quoted strings automatically.
# Splitting the address across lines makes it easier to read.
data_url = (
    'https://raw.githubusercontent.com/joshuakalla/data_science_campaigns_26/'
    'df35e89ef09877aee3f2de4ce49a1ec654520e7f/weeks/wk05_power_and_nulls/'
    'data/canvassing_pilot.csv'
)

# Read the online CSV into a pandas DataFrame: a table named df.
# This step requires an internet connection.
df = pd.read_csv(data_url)

# .shape returns (number of rows, number of columns).
# .head() displays the first five rows so we can inspect the column names
# and see how assignment and turnout are recorded.
print('Rows and columns:', df.shape)
df.head()

Rows and columns: (400, 3)


,household_id,treatment,turned_out
0,1,1,1
1,2,0,0
2,3,0,0
3,4,0,0
4,5,1,1


In [2]:
# Work from the inside out:
# 1. df['treatment'] == 1 marks treatment rows with True.
# 2. df[...] keeps those rows.
# 3. ['turned_out'] selects their turnout values.
# 4. .mean() averages the 0/1 values, giving the proportion who voted.
treated_mean = df[df['treatment'] == 1]['turned_out'].mean()

# Repeat the same steps for the control group, whose assignment equals 0.
control_mean = df[df['treatment'] == 0]['turned_out'].mean()

# Always subtract control from treatment. A positive ATE means higher
# turnout among households assigned to a canvass attempt.
observed_ate = treated_mean - control_mean

# For a binary outcome, p * (1 - p) describes outcome variability.
# Dividing by the group's size gives the estimated variance of its mean.
# Add the two variances for independent groups, then take the square root
# to obtain the standard error of the difference. Each group has 200 rows.
se = np.sqrt(treated_mean * (1 - treated_mean) / 200
             + control_mean * (1 - control_mean) / 200)

# Subtract and add the approximate 95% margin of error.
# This assigns the lower endpoint to lo and the upper endpoint to hi.
lo, hi = observed_ate - 1.96 * se, observed_ate + 1.96 * se

# The f before a string inserts calculated values inside {...}.
# Multiplying turnout by 100 gives percentages; multiplying differences,
# SEs, and interval endpoints by 100 gives percentage points (pp).
# :.1f displays one decimal place; :+.1f also displays the sign.
print(f'Control: {100 * control_mean:.1f}%; treatment: {100 * treated_mean:.1f}%')
print(f'ATE: {100 * observed_ate:+.1f} pp; SE: {100 * se:.1f} pp')
print(f'95% CI: [{100 * lo:+.1f}, {100 * hi:+.1f}] pp')

Control: 42.0%; treatment: 43.5%
ATE: +1.5 pp; SE: 4.9 pp
95% CI: [-8.2, +11.2] pp


In [3]:
# In this formula, the outcome goes to the left of ~ and the explanatory
# variable goes to the right. data=df tells the model where to find them.
# ols() sets up ordinary least squares; .fit() estimates the coefficients.
# An intercept is included automatically. With a 0/1 treatment variable,
# its coefficient equals treatment mean minus control mean.
pilot_model = smf.ols('turned_out ~ treatment', data=df).fit()

# Display the fitted model. Focus on the treatment row's coefficient,
# standard error, p-value, and confidence interval. The coefficient and
# interval are in proportions: 0.015 corresponds to +1.5 pp.
pilot_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             turned_out   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                   0.09149
Date:                Thu, 01 Oct 2026   Prob (F-statistic):              0.762
Time:                        16:12:46   Log-Likelihood:                -286.02
No. Observations:                 400   AIC:                             576.0
Df Residuals:                     398   BIC:                             584.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4200      0.035     11.978      0.000       0.351       0.489
treatment      0.0150      0.050      0.302      0.762      -0.082       0.112
==============================================================================
Omnibus:                     1849.723   Durbin-Watson:                   2.104
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               66.730
Skew:                           0.293   Prob(JB):                     3.23e-15
Kurtosis:                       1.087   Cond. No.                         2.62
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Read the **treatment row**: `coef` is the ATE; `std err` is its SE;
<code>P&gt;&#124;t&#124;</code> is the p-value; `[0.025, 0.975]` gives the CI endpoints.
Multiply coefficients, SEs, and endpoints by 100 for pp, but not the p-value.
The regression's t-based interval is very close to the calculation above.

### Randomization inference

Keep outcomes fixed and shuffle assignment labels under the **sharp null
of no effect on any household**. `.append()` stores each gap; the final
mean counts the fraction at least as extreme in either direction.

In [4]:
# Start a reproducible sequence of random shuffles.
# Set the seed once before the loop so successive shuffles can differ.
np.random.seed(42)

# An empty list will hold one treatment-control gap per shuffle.
fake_ates = []

# range(1000) supplies 0 through 999. The indented block runs 1,000 times.
for i in range(1000):
    # Under the sharp null, every observed turnout value stays fixed.
    # .sample(frac=1) reorders all assignment labels without replacement,
    # preserving the number of treated and control households.
    # .values removes the old row index, so labels are used in their new order.
    shuffled = df['treatment'].sample(frac=1).values

    # Select outcomes according to the NEW labels for this shuffle.
    fake_treated = df['turned_out'][shuffled == 1]
    fake_control = df['turned_out'][shuffled == 0]
    fake_ate = fake_treated.mean() - fake_control.mean()

    # Save this gap before the next iteration replaces fake_ate.
    fake_ates.append(fake_ate)

# The list comprehension checks every stored gap, calling it f in turn.
# abs() removes the sign: large gaps in either direction count as extreme.
# >= includes ties with the observed gap. Each comparison is True or False.
# np.mean() counts True as 1 and False as 0, giving the extreme fraction.
# This is a Monte Carlo estimate of the two-sided RI p-value.
p_value = np.mean([abs(f) >= abs(observed_ate) for f in fake_ates])
print(f'RI p-value: {p_value:.3f}')  # :.3f displays three decimal places.

RI p-value: 0.841


The empirical RI p-value is about **0.84**, versus **0.76** from
regression. The procedures differ; neither is significant here.

## 3. A world where canvassing works

Assume control turnout is **0.42** and the true effect is **0.02**.
`np.random.binomial(1, p, size=n)` draws `n` independent 0/1 outcomes;
`.mean()` gives turnout and `.sum()` counts voters. A seed makes the
random sequence reproducible. Here is one simulated experiment.

In [5]:
# These are assumptions for a simulated experiment, not fitted values.
np.random.seed(2028)
n_per_group = 200            # 200 in each group means 400 households total.
true_control_rate = 0.42     # Each control household has a 42% turnout chance.
true_effect = 0.02           # A difference of 0.02 is +2 percentage points.
true_treatment_rate = true_control_rate + true_effect  # 0.44, or 44%.

# binomial(1, p, size=n) draws n independent 0/1 outcomes.
# The first argument, 1, means one trial per household; p is its chance of 1.
# The realized turnout rate can differ from p because of random variation.
control_outcomes = np.random.binomial(1, true_control_rate, size=n_per_group)
treatment_outcomes = np.random.binomial(1, true_treatment_rate, size=n_per_group)

# Average each group's outcomes and calculate this experiment's estimate.
# The true effect is fixed at +2 pp; fake_ate varies across experiments.
fake_ate = treatment_outcomes.mean() - control_outcomes.mean()

# For 0/1 outcomes, .sum() counts turnout successes and .mean() gives a rate.
print(f'Control voters: {control_outcomes.sum()} of {n_per_group}')
print(f'Treatment voters: {treatment_outcomes.sum()} of {n_per_group}')
print(f'True effect: {100 * true_effect:+.1f} pp')
print(f'ATE in this experiment: {100 * fake_ate:+.1f} pp')

Control voters: 86 of 200
Treatment voters: 91 of 200
True effect: +2.0 pp
ATE in this experiment: +2.5 pp


### Repeat, test, count

The **outer loop** creates 1,000 new experiments where the effect is +2 pp.
The **inner loop** tests each one using 200 RI shuffles. Pooling and
re-splitting outcomes is equivalent to reassigning fixed outcomes under
the null. Power is the fraction of **experiments**, not shuffles, with
`p < 0.05`. Reset `ri_count` inside the outer loop; reset `detected` before it.

In [6]:
# Keep the design fixed while repeatedly generating NEW experiments.
np.random.seed(42)
n_per_group = 200
true_control_rate = 0.42
true_treatment_rate = 0.44  # The true effect is +2 pp in every experiment.
n_experiments = 1000        # How many datasets to generate for estimating power.
n_shuffles = 200            # How many null reassignments to test each dataset.

# This counter records experiments that reject the null.
# Initialize it once so it can accumulate across the outer loop.
detected = 0

for i in range(n_experiments):
    # OUTER LOOP: draw a new dataset and calculate its observed gap.
    control = np.random.binomial(1, true_control_rate, size=n_per_group)
    treatment = np.random.binomial(1, true_treatment_rate, size=n_per_group)
    ate = treatment.mean() - control.mean()

    # Join the two arrays into one pool of outcomes for this experiment.
    # These outcomes stay fixed throughout its inner randomization test.
    all_outcomes = np.concatenate([control, treatment])

    # Reset this counter for EACH experiment: its p-value starts from zero
    # extreme shuffles, regardless of what happened in the previous dataset.
    ri_count = 0
    for j in range(n_shuffles):
        # INNER LOOP: shuffle the pooled outcomes and split into equal groups.
        # This is equivalent to randomly reassigning labels to fixed outcomes.
        shuffled = np.random.permutation(all_outcomes)

        # [:n_per_group] selects the first group; [n_per_group:] the rest.
        # The stop position is excluded. We compare absolute gaps below,
        # so reversing the subtraction order does not change this test.
        ri_ate = shuffled[:n_per_group].mean() - shuffled[n_per_group:].mean()
        if abs(ri_ate) >= abs(ate):
            ri_count += 1  # += 1 increases the counter by one.

    # Back outside the inner loop: calculate ONE p-value for this dataset.
    # Divide by shuffles, because ri_count counts extreme reassignments.
    p = ri_count / n_shuffles

    # A significant experiment contributes one to the detection counter.
    if p < 0.05:
        detected += 1

# Back outside BOTH loops: the detection fraction estimates power.
# Divide by experiments, because detected counts significant datasets.
power_400 = detected / n_experiments
print(f'Total households: {2 * n_per_group}')
print(f'Experiments detecting the effect: {detected} / {n_experiments}')
print(f'Power: {power_400:.1%}')  # :.1% formats 0.047 as 4.7%.

# Since an effect exists in this simulated world, a non-detection is a
# false negative. Its frequency is one minus the detection frequency.
print(f'Fraction missing the effect: {1 - power_400:.1%}')

Total households: 400
Experiments detecting the effect: 47 / 1000
Power: 4.7%
Fraction missing the effect: 95.3%


The result need not be exactly 6%: the seed and finite number of shuffles
affect it.

## 4. Size, MDE, and cost

For speed, the sample-size sweep uses the approximate rule
`abs(ate) > 1.96 * SE`, with planning turnout 0.43. It need not match RI
exactly. We also use the **2.8 × SE** rule to report approximate MDEs.
`n_total // 2` puts half the households in each group.

In [7]:
# Compare several designs, always assuming a true +2 pp effect.
np.random.seed(42)
sample_sizes = [400, 1000, 2000, 5000, 8000, 10000, 15000, 20000]
power_rows = []  # Store one summary row for each total sample size.

for n_total in sample_sizes:
    # // performs integer division. Every total here is even, so this
    # gives two equally sized groups with no households left over.
    n_per = n_total // 2

    # Approximate SE using the fixed planning turnout rate of 0.43.
    # 0.57 is 1 - 0.43; the factor 2 accounts for the two equal groups.
    # A larger n_per reduces the SE and the threshold for significance.
    planning_se = np.sqrt(2 * 0.43 * 0.57 / n_per)
    threshold = 1.96 * planning_se
    detected = 0  # Start a fresh count for this sample size.

    for i in range(1000):
        # Generate one experiment at the CURRENT group size.
        control = np.random.binomial(1, 0.42, size=n_per)
        treatment = np.random.binomial(1, 0.44, size=n_per)
        ate = treatment.mean() - control.mean()

        # Use the fast approximate test instead of an inner shuffle loop.
        # abs() counts sufficiently large estimates in either direction.
        if abs(ate) > threshold:
            detected += 1

    # A dictionary pairs column names with this design's results.
    # Power is a proportion. MDE is expressed in percentage points:
    # multiply the approximate 80%-power MDE, 2.8 * SE, by 100.
    power_rows.append({'n_total': n_total, 'power': detected / 1000,
                       'MDE_pp': 100 * 2.8 * planning_se})

# Convert the list of dictionaries into a table with one row per design.
# .round(3) displays three decimal places; power_table keeps full precision.
power_table = pd.DataFrame(power_rows)
power_table.round(3)

,n_total,power,MDE_pp
0,400,0.075,13.862
1,1000,0.097,8.767
2,2000,0.171,6.199
3,5000,0.275,3.921
4,8000,0.428,3.100
5,10000,0.516,2.772
6,15000,0.694,2.264
7,20000,0.817,1.960


### From sample size to cost

Suppose the average canvassing cost is **20 USD per completed conversation**,
including the effort spent on unanswered doors. A **30% contact rate** means
100 attempted doors produce about 30 conversations. Those attempts therefore
cost about **30 × 20 = 600 USD**, or **600 / 100 = 6 USD per attempted door**.

With equal assignment, half the sampled households receive a canvass attempt
and half are controls. Multiply the number assigned to treatment by the
per-door cost:

`study_cost = (n_total // 2) * 6`

Here `n_total // 2` is the treatment-group size. For example, 2,000 households
in total means 1,000 treatment attempts, costing **6,000 USD**. Outcome
measurement and other study costs outside the canvassing rate need separate
budget lines.

## 5. Exercises

Complete the five exercises below. When finished, restart and run all cells.
Check that both coding exercises produce output and that the written answers
are filled in; comment-only cells can run while unfinished.

### Exercise 1

A volunteer reports success in **15 of 20** conversations. Assume independent
conversations with success probability **0.50**. Set seed 42. Draw 20 outcomes,
use `.sum()`, and store the count in a list. Repeat **1,000** times, then
calculate the fraction with **at least 15** successes. Name it
`fraction_significant` and print it. Use one loop, without an inner RI test.

**Check:** the fraction should be below 5%. Explain whether 15/20 is unusual
under this model. Reported successes alone do not establish a causal turnout effect.

In [8]:
# Write your simulation below, using the steps in the exercise prompt.
# Store one SUCCESS COUNT per repetition, rather than all individual outcomes.
# Keep the final fraction calculation outside the loop, after the list is full.
# Check that the final result is between 0 and 1.
# End by printing fraction_significant and write your interpretation below.

**Your interpretation:**

### Exercise 2

Use `power_table` to identify the total sample giving about 80% power for
+2 pp and compare its MDE with the pilot's. Set `chosen_n_total`, calculate
`chosen_n_treated`, and print **`study_cost`** using 6 USD per attempted door.

In [10]:
# Read the total sample size from power_table before writing the calculation.
# Keep separate variables for all households and treatment households.
# Track units: households receiving an attempt times USD per attempted door.
# End by printing study_cost with units, then interpret your chosen design below.

**Your sample-size and MDE interpretation:**

### Exercise 3

Two equally credible trials of election-day reminder calls report:

| Trial | Estimate (pp) | 95% CI (pp) |
|---|---:|---|
| A | −0.6 | [−5.3, +4.1] |
| B | +0.6 | [−0.5, +1.7] |

Both teams need a gain of **at least +2.6 pp** to justify the program.
Write a two-sentence decision note: which trial leaves that gain plausible,
and which provides evidence against a gain that large? Avoid claiming zero effect.

Trial A could expand from **750 to 6,750 households per group**. With the
same design and similar turnout, by what factor would its CI width change?
Explain without recomputing the endpoints.

**Your answer:**

### Exercise 4

For a fictional two-arm registration-reminder trial, an analyst writes:

```python
# All quantities in this note use percentage-point units.
se_pp = 0.85
# Inspect whether each formula matches the name on its left.
mde_pp = 1.96 * se_pp
significance_cutoff_pp = 2.8 * se_pp
```

The target is 80% power with a two-sided 5% test. Correct the two labels or
formulas and explain what each quantity means. Then place an observed
**+2.05 pp** estimate relative to both corrected cutoffs: what would the
test conclude, and why is that compatible with the MDE?

**Your answer:**

### Exercise 5

In a toy simulation, six small trials share a true effect of **+1.1 pp**
and each reports **SE = 1.9 pp**. Their estimates, in pp, are:

`−2.7, −0.8, +0.3, +1.7, +3.6, +4.5`

A report keeps only positive estimates exceeding **1.96 × SE**.
Identify the retained estimates and compare their mean with the mean of
all six. Explain how the reporting rule changes the apparent effect even
though the true effect is unchanged. What would you ask the report to show?

**Your answer:**